### ISSUE TITLE
simplified czech layout second row misaligned

### ISSUE THREAD

**Issue by cptMikky** [AnySoftKeyboard/LanguagePack#266]
The simplified czech qwerty layout has misaligned second row (the "ASDF" row).

The row should be aligned to the center but it is instead aligned to the left with a hole on the right (where the "Ů" symbol is on the full keyboard layout).

(Lineage 15.1 and Lineage 16, latest packages from f-droid)

**Comment by cptMikky** [comment]
Although I don't really understand why this is happening as https://github.com/AnySoftKeyboard/LanguagePack/blob/Czech/src/main/res/xml/cs_qwerty_simple.xml (line 20) looks fine.

Maybe a bug somewhere?

**Comment by ArenaL5** [comment]
F-Droid's most current download is from July 2018. Your bug was already fixed.

Line 20 in the code for the APK in F-Droid reads this:
```
    <Key android:codes="a" android:popupCharacters="\@áäα" ask:hintLabel="\@" android:keyEdgeFlags="left" />
```

Without an `android:horizontalGap="5%p"`, the second row is aligned to the left (or, presumably, to the right if you were using an RTL layout).

I don't speak Czech, but I've just published code for a similar issue in another keyboard. I'll bundle a quick fix for #267 so that hopefully, ASK's team will push both fixes to F-Droid.

**Comment by menny** [comment]
Just to be clear, updating FDroid packages **does not** require any action on my part. Anyone can do that.

The file https://gitlab.com/fdroid/fdroiddata/blob/master/metadata%2Forg.herrlado.ask.languagepack.czech.yml needs to be updated and point to the latest commit.

Unfortunately, I have no idea how to do it with the current repo layout.
If someone figures that out, it would be great.

**Comment by ArenaL5** [comment]
One still needs write access to the repo to edit that file in GitLab.

[The F-Droid metadata format specifications](https://f-droid.org/en/docs/Build_Metadata_Reference/) should help, but this... is going to be difficult.

Is there any way to build APKs for a single language, rather than for every one of them? I get the impression that it would help F-Droid to find every version of APKs. (Or at the very least, would be much faster for them.)

Or maybe you can try adding a code like this. I have no way to test it:

```
  - versionName: 2.0.3
    versionCode: 261
    commit: ec2aa6498329880127a157d2442453cce75ea854
    gradle:
      - yes
    forceversion: true
    forcevercode: true
    prebuild: sed -i -e 's/versionCode versionData.versionCode/versionCode 261/' build.gradle
    preassemble:
      - :makeDictionary
    output: add_ons_apks/release/add-on--languages-czech-apk-1.apk
```

**Comment by menny** [comment]
@ArenaL5 if we just want to build czech, then we can call `./gradlew :languages:czech:apk:assembleRelease`
This will output an APK at `add_ons_apks/release/add-on--languages-czech-apk-1.apk`
The problem is that it will be versioned with `1`

**Comment by menny** [comment]
I think I can add another command-line argument that will force the version.
Let me work on that.

@ArenaL5 thanks for finding the documentation.

**Comment by menny** [comment]
Okay, I've bumped the auto-version plugin to allow forcing build-counter value.

I think we should use this:
```
  - versionName: 4.0.634
    versionCode: 2234
    commit: 1e7ab8bc4a39d71ac67dbae9555850b087b7a289
    gradle:
      - yes
    gradleprops: forceVersionBuildCount=1234
    output: add_ons_apks/release/add-on--languages-czech-apk-2234.apk
```

**Comment by menny** [comment]
I took the values from the last deploy to Play Store in circle-ci: https://app.circleci.com/jobs/github/AnySoftKeyboard/LanguagePack/1234

**Comment by menny** [comment]
I couldn't locate how to only build the Czech APK, but since we specify the `output`, I don't think there should be a problem with building everything.

If this looks sane, I can try to push this to fdroid.

**Comment by ArenaL5** [comment]
Glad I could help. The lines I can understand look sane to me.

F-Droid documentation says that `output` is supposed to point to an unsigned APK. I think that your gradle script defaults to unsigned if it can't find any keys, so it should work when they build ASK.

If not, maybe swapping `release` for `debug` in the output path would work, but I doubt it.

**Comment by menny** [comment]
we can try out this and see if it works.

**Comment by menny** [comment]
https://gitlab.com/fdroid/fdroiddata/merge_requests/5752

**Comment by ArenaL5** [comment]
Line 54 needs an extra space of indentation.

...yes, I believe that's the only error. :rofl:

**Comment by menny** [comment]
```
  - versionName: 4.0.634
    versionCode: 2234
    commit: 1e7ab8bc4a39d71ac67dbae9555850b087b7a289
    gradle:
      - yes
    output: add_ons_apks/release/add-on--languages-czech-apk-2234.apk
    gradleprops:
      - forceVersionBuildCount=1234

```

**Comment by menny** [comment]
had to create another MR: https://gitlab.com/fdroid/fdroiddata/merge_requests/5753.

**Comment by ArenaL5** [comment]
Okay. There's hope, but we're not quite there yet. Now it says at GitLab that `fdroid build` is not compiling correctly.

I tried to learn a little better how do the F-Droid builds work, and after much trial and error I found a way to run `fdroid build` on my computer and get an error log. (It's not a problem with the metadata file, or with having several projects in one repo.) This is the progress I made so far:

Setting up `fdroid build`
----

You may want to skip every `fdroid bulid` command except the last, as they'll probably raise an error. I'm writing the full process just in case paths are sligthly different on your computer.

It should be possible to run `fdroid build` from the source, including the metadata file in the root of the project under `circle.yml` or `.circle.yml`. But it couldn't find the app ID and I couldn't find a way to insert it, so I decided to use the other method and use only F-Droid's metadata files.

1. `sudo apt install fdroidserver`, it will probably ask you to install 50 packages worth of dependencies, that's normal.
2. Choose a folder, then `git clone https://gitlab.com/fdroid/fdroiddata.git`. That downloads the metadata for the full main F-Droid repo (not the actual source code or APKs, so it should only take a couple minutes)
3. `cd fdroiddata/metadata`
4. `fdroid build -t org.herrlado.ask.languagepack.czech:260` (The `-t` option is intended for "test builds", not ready for publishing yet - I think it just moves the finished APK to a different folder. The number after the colon is the `versionCode`)

At this point, the build process will probably fail with this kind of error. (Translated from Spanish, actual wording may differ:)

```CRITICAL: 'bc1qu482c0tngkcvx3q7mrm8zmuldrh2f2lrh26ym0' is not a valid Bitcoin address at org.projectmaxs.transport.xmpp. Regex pattern: ^[a-zA-Z0-9]{27,34}$```

That seems like a problem they have upstream, and maybe I'll look into it later some time after this is sorted out. The apps that raise that error aren't related to yours, so I just deleted their metadata files and tried again:

5. `rm -R metadata/org.projectmaxs.* metadata/net.fabiszewski.ulogger*`
6. `fdroid build -t org.herrlado.ask.languagepack.czech:260`

`fdroid build` now outputs some promising information, ending with this error line:

```ERROR: Could not build app org.herrlado.ask.languagepack.czech: OSError while trying to execute /usr/lib/python3/dist-packages/gradlew-fdroid clean: [Errno 2] No such file or directory: '/usr/lib/python3/dist-packages/gradlew-fdroid': '/usr/lib/python3/dist-packages/gradlew-fdroid'```

After several attempts, I found how to download `gradlew-fdroid`:

7. `cd ../..`, we'll git-clone another project and we don't want both of them to mix.
8. `git clone https://gitlab.com/fdroid/fdroidserver.git`
9. `sudo cp fdroidserver/gradlew-fdroid /usr/lib/python3/dist-packages/gradlew-fdroid`
10. `cd fdroiddata`
11. **Finally**, `fdroid build -t org.herrlado.ask.languagepack.czech:260` should work this time, and compile the version that's currently on F-Droid.

Every time I tried to build it raised this warning, but it didn't affect the build. I simply ignored it.

(translated from Spanish)

```WARNING: 'config.py' not found, default values will be used```

Actual cause of error
-----

At this point, still from the `fdroiddata` folder, I overwrote `metadata/org.herrlado.ask.languagepack.czech.yml` with the file you're trying to merge.

I double-checked `fdroid build -t org.herrlado.ask.languagepack.czech:260` still worked, and tried to compile the new version, with `fdroid build -t org.herrlado.ask.languagepack.czech:2234`.

This is the log. You can ignore the sentences in Spanish, they just parrot the same information (the first line is the warning for not finding `config.py`).

```
WARNING: No se encontró 'config.py', se usarán los valores predeterminados.
INFO: Using git version 2.20.1
INFO: Building version 4.0.634 (2234) of org.herrlado.ask.languagepack.czech
INFO: Getting source for revision 1e7ab8bc4a39d71ac67dbae9555850b087b7a289
INFO: Creating local.properties file at build/org.herrlado.ask.languagepack.czech/local.properties
INFO: Cleaning Gradle project...
INFO: Scanning source for common problems...
ERROR: Found unknown maven repo 'https://dl.bintray.com/novoda-oss/snapshots/' at build.gradle
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MainClass.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MakeBinaryDictionary$1.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MergeWordsListTask.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MakeBinaryDictionary$CharNode.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/XmlWriter.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/BinaryDictionaryResourceNormalizer.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MergeWordsListTask$MySaxHandler.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MakeDictionaryPlugin.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/GenerateWordsListTask.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/WordWithCount.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/WordListWriter.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MakeDictionaryTask.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/Parser.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/MakeBinaryDictionary.class
ERROR: Found Java compiled class at buildSrc/build/classes/java/main/GenerateWordsListFromAOSPTask.class
WARNING: Found JAR file at buildSrc/build/libs/buildSrc.jar
WARNING: Found possible binary at languages/armenian2/pack/src/main/res/raw/armenian2_words_2.dict
WARNING: Found possible binary at languages/armenian2/pack/src/main/res/raw/armenian2_words_3.dict
WARNING: Found possible binary at languages/armenian2/pack/src/main/res/raw/armenian2_words_4.dict
WARNING: Found possible binary at languages/armenian2/pack/src/main/res/raw/armenian2_words_1.dict
INFO: Removing gradle-wrapper.jar at gradle/wrapper/gradle-wrapper.jar
WARNING: Found possible binary at quicktexts/bbcodes/apk/flag/flag.png
ERROR: Could not build app org.herrlado.ask.languagepack.czech: No se puede construir debido a los errores 16 mientras se comprobaba
INFO: Build for app org.herrlado.ask.languagepack.czech failed:
No se puede construir debido a los errores 16 mientras se comprobaba
INFO: Terminado
INFO: 1 build failed
```

**Comment by ArenaL5** [comment]
**TL;DR:** So apparently `fdroid build` doesn't like having pre-compiled Java files at `buildSrc/build/classes/java/main/*.class` (were thosed compiled from the main repo?) and, while reading this line at `build,gradle`:

```        maven { url 'https://dl.bintray.com/novoda-oss/snapshots/' }```

raises this error:

```Found unknown maven repo 'https://dl.bintray.com/novoda-oss/snapshots/' at build.gradle```

**Comment by menny** [comment]
thanks!

**Comment by menny** [comment]
I updated the MR with to the latest commit
